Move Float16 print tests to their own file and test exhaustively

stephentyrone opened the pull request:
We can easily test all 2**16 values, so let's do it. Also now _Float16 is properly supported in clang, so we can pass arguments to CPP that way, which lets us get snan right on more platforms and just generally make things cleaner.

stephentyrone committed:
Move testing of Float16 printing to a separate file and test exhaustively.

We can easily test all 2**16 values, so let's do it. Also now _Float16 is properly supported in clang, so we can pass arguments to CPP that way, which lets us get snan right on more platforms.

stephentyrone commented:
@swift-ci test

mikeash commented on test/stdlib/PrintFloat16.swift:
```diff
@@ -0,0 +1,163 @@
+//    } else {
+//      setlocale(LC_ALL, "")
+//    }
+//  }
```
Should this be uncommented, or perhaps deleted?

stephentyrone committed:
Remove some unused locale code

And fixup availability for x86 that was copied from another branch.

stephentyrone committed:
Condition _Float16 usage in C on FLT16_MAX.

stephentyrone commented:
@swift-ci test

stephentyrone committed:
Move check for FLT16_MAX to header.

stephentyrone commented:
@swift-ci smoke test

stephentyrone committed:
Use swift_dtoa_optimal_binary16_p instead of the non-pointer version.

Apparently clang/windows (sigh) doesn't define FLT16_MAX in <float.h>,
so we have to keep using the _p interface instead.

stephentyrone commented:
@swift-ci smoke test
